Discord enhancements

cheesegrits opened the pull request:
The main features of this PR are ...

Split the '!schedule' command into '!sessions' (full "wall of text" individual session calendar, restricted to one week ahead) and '!schedule' (abbreviated, one per race calendar)

Added role mentioning.  If the optional DiscordRoleID is set, that role will be mentioned in all Discord notifications (meaning anyone with that role will get pinged).  Additionally, if the optional 'DiscordRoleCommand' is also set, we will attempt to add/remove that role for users, when they issue the "!whatever" command verb - this requires that the bot has "Manage Roles" permission.

What we do with this feature is, we have a 'schedule-announcements' channel, which only the bot and mods can post in.  This code mentions the role in all posts. Users can add/remove the role easily with "!notify".  This provides uncluttered, easy to use notifications, with/without pings.

General tidy up of message formats, added server name to messages, switch to logrus.WithError(), etc.

cheesegrits commented:
PS, no hurry on merging it.  I've been running it live for about a day, no doubt I'll do some more tweaking over the next few days.  Just wanted to get it on the radar for either next release, or when race weekend dust has settled.

cj123 commented on discord.go:
```diff
@@ -96,35 +98,202 @@ func (dm *DiscordManager) SaveServerOptions(oldServerOpts *GlobalServerConfig, n
+	}
+
+	// w00t!
+	return fmt.Sprintf("The %s role has been assigned, you will now get pinged with notifications.  Type the command again to remove it."), nil
```
this line is causing a build failure. You're using `%s` in the fmt string, but not providing an argument for it :)

cheesegrits replied:
Yeah, I noticed and fixed that the first time I built it, just haven't committed yet.

cj123 commented on discord.go:
```diff
@@ -2,12 +2,14 @@ package servermanager
 	"github.com/bwmarrin/discordgo"
+	"github.com/google/uuid"
 	"github.com/sirupsen/logrus"
+	"net/url"
```
can you group the stdlib imports at the top? i know i'm being picky haha.

i tend to do:

* stdlib imports at the top
* a gap, then sub-package imports
* another gap (if necessary) and then third party imports

so like
```go
import (
    "errors"
    "url"

    "github.com/cj123/assetto-server-manager/pkg/udp"

    "github.com/google/uuid"
    "github.com/sirupsen/logrus"
)
```

cheesegrits replied:
Okie dokie

cj123 commented on discord.go:
```diff
@@ -96,35 +98,202 @@ func (dm *DiscordManager) SaveServerOptions(oldServerOpts *GlobalServerConfig, n
+
+	scheduled = append(scheduled, recurring...)
+
+	var msg = fmt.Sprintf("\nUpcoming events on server %s\n\n", serverOpts.Name)
```
again i'm being picky:

please can you only use `var` to define a variable inside a function when it is not being assigned immediately. for every other variable, use `:=`.

I like that `var` signifies that the variable exists but doesn't have a value yet. just a little readability thing again :)

cheesegrits replied:
I'd originally created that as an empty string, then at some point came back and added that initial line to it, forgot to do the := thing

cj123 commented on discord.go:
```diff
@@ -96,35 +98,202 @@ func (dm *DiscordManager) SaveServerOptions(oldServerOpts *GlobalServerConfig, n
+	}
+
+	if err != nil {
+		logrus.WithError(err).Errorf("Error during handling of Discord command")
```
if the err here is non-nil, do you want the ChannelMessageSend below to occur?

cheesegrits replied:
Yes, because the specific command handler will have returned some user feedback in msg, like "An error occurred trying to assign you that role, please try again later", rather than just silently failing.  So I log the error for the admin to see, and send the user feedback.

cj123 commented on discord.go:
```diff
@@ -96,35 +98,202 @@ func (dm *DiscordManager) SaveServerOptions(oldServerOpts *GlobalServerConfig, n
+
+	for _, scheduledEvent := range scheduled {
+		raceSetup := scheduledEvent.GetRaceSetup()
+		trackInfo := trackInfo(raceSetup.Track, raceSetup.TrackLayout)
```
`trackInfo` i believe can return a nil value (e.g. if there is no ui_track.json). there's a new helper function now in master called 'trackSummary' which i think might be more useful. it will always return something :)

cheesegrits replied:
Ah yes, I discovered that nil thing elsewhere in the code recently, and did indeed switch to using trackSummary.  Forgot to check for other uses of trackInfo().  Will fix.

cj123 commented on discord.go:
```diff
@@ -96,35 +98,202 @@ func (dm *DiscordManager) SaveServerOptions(oldServerOpts *GlobalServerConfig, n
+		return fmt.Sprintf("A server error occurred trying to assign you the %s role, try again later", roleName), nil
+	}
+
+	// w00t!
```
:tada: :grin:

cheesegrits replied:
There was much celebrating when I finally got to that line in testing.  :)

cheesegrits commented:
@cj123 meh, I haz issue.

I'm trying to convert NotificationReminderTimer from an int to a string, so we can specify multiple reminder times in a comma separated list, like 120,15.  That was the first request I got from several folk this week, when we debuted this system.  Could they have another reminder a couple of hours ahead as well as the 15m one.

I had thought about doing that to start with, but decided to leave it for later.  Only now I'm in migrate hell.  If I change the definition in config.ini.go to string, then the first attempt to load the config blows up with an unmarshall error.  But I obviously can't change the type in a migrate function.

I just don't see a way out of this one.  Looked through all the existing migrate functions, but no precedent for changing a type.  What I've done for now is just change the name to NotificationReminderTimers (plural) ... which means the original option just gets lost, and has to be re-entered in the Options.

Tbh, I doubt anyone but you and I are using this atm, so this may be OK.  But obviously if there is a way of doing it that preserves the existing value, that'd be better.

cj123 commented:
Ah, yeah there's no real precedent for that. I think what I would do is change the name of the value in the struct (if possible), change the type to a string, but keep the old one as well and mark it as deprecated and hide it from the form.

Then you can write a migration which reads the old one and sets it to the new one.

cheesegrits commented:
OK, then at a later date, a second migration to remove the old one.  I shoulda just set the original to be a string, even though it was for a single it.  Insufficient forethought.

cj123 commented:
> Almost there. Got the migration working, with a new option, leaving the old one and hiding it. There's a code review request, where I had to add a show="hidden" type, as it didn't seem like any of the existing show options would just straight up hide regardless.
> 
> Couple of changes in race manager for the new multiple reminders, and in the dummyNotificationManager testing.

`show:"-"` should do what you described. I designed it following the same kind of logic as go's json package (which uses `json:"-"` to signify ignoring a field). I realise this probably isn't too clear to someone new to go!

cheesegrits commented:
BTW, apologies for the size of this PR.  I usually try and keep PRs to small, individual little bug fix or single feature branches, but this week has been the first week we started using this stuff for realz in our league, so I had a bunch of stuff to add/fix.

cheesegrits commented:
@cj123 I just merged master again, for the 1.5.1 stuff.  Should be good to go.

cheesegrits commented:
Change notes:

Split the '!schedule' command into '!sessions' (full "wall of text" individual session calendar, restricted to one week ahead) and '!schedule' (abbreviated, one per race calendar).  This still needs work, as can easily exceed Discord's max msg length.

Added role mentioning. If the optional DiscordRoleID is set, that role will be mentioned in all Discord notifications (meaning anyone with that role will get pinged).   Additionally, if the optional 'DiscordRoleCommand' is also set, we will attempt to add/remove that role for users, when they issue the "!whatever" command verb - this requires that the bot has "Manage Roles" permission.

Changed NotificationReminderTimer to NotificationReminderTimers (plural), to support comma separated multiple timers (like "90,15" for two reminders at 90 and 15 minutes).

Added option to disable notifications when a race is scheduled.

Added notification for scheduled races being cancelled.

Improved formatting of Discord messages, everything is now an embed (except the role mention, which has to be part of the basic message).
